### PR TITLE
[bug] MatrixType bug fix: Fix error with BLS

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: https://github.com/taichi-dev/taichi/issues/5819

### Brief Summary
Fixed a set of issues to make BLS tests work.
1. Modified GroupedNDRange generator to directly yield `Expr with TensorType` instead of `_IntermediateMatrix` when `real_matrix=True`
2. Added support for `rescale_index()` to handle `Expr with TensorType`
3. Added scalarization for `indices` of SNode ops

### CODE REVIEW

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

File: taichi/transforms/scalarize.cpp [hunk 1]
```diff
@@ -9,6 +9,23 @@
 
 namespace taichi::lang {
 
+template <typename T>
+static std::vector<Stmt *> get_scalarized_indices(T *stmt) {
```
Comment (strongoier) [inline taichi/transforms/scalarize.cpp:13]: Why do we have to scalarize indices here? Have they been already scalarized at https://github.com/taichi-dev/taichi/blob/4f7f2f568d655e6fc507652f0d669c4aa077b75e/python/taichi/lang/impl.py#L166-L179

File: python/taichi/lang/snode.py [hunk 2]
```diff
@@ -365,17 +395,25 @@ def rescale_index(a, b, I):
         I = matrix.Vector(I)
```
Comment (strongoier) [inline python/taichi/lang/snode.py:395]:
```suggestion
        n = len(I)
```

File: python/taichi/lang/snode.py [hunk 3]
```diff
@@ -365,17 +395,25 @@ def rescale_index(a, b, I):
...
-        ), "The third argument must be an index (list or ti.Vector)"
-    entries = [I(i) for i in range(I.n)]
-    for n in range(min(I.n, min(len(a.shape), len(b.shape)))):
-        if a.shape[n] > b.shape[n]:
-            entries[n] = I(n) // (a.shape[n] // b.shape[n])
-        if a.shape[n] < b.shape[n]:
-            entries[n] = I(n) * (b.shape[n] // a.shape[n])
-    return matrix.Vector(entries)
+            I, (list, expr.Expr, matrix.Matrix)
+        ), "The third argument must be an index (list, ti.Vector, or Expr with TensorType)"
```
Comment (strongoier) [inline python/taichi/lang/snode.py:399]:
```suggestion
            I, (expr.Expr, matrix.Matrix)
        ), "The third argument must be an index (list, ti.Vector, or Expr with TensorType)"
        n = I.n
```

File: python/taichi/lang/snode.py [hunk 4]
```diff
@@ -365,17 +395,25 @@ def rescale_index(a, b, I):
...
+    from taichi.lang.kernel_impl import pyfunc  # pylint: disable=C0415
+
+    @pyfunc
+    def _rescale_index():
+        entries = [I[i] for i in range(I.n)]
+        for n in impl.static(range(min(I.n, min(len(a.shape), len(b.shape))))):
+            if a.shape[n] > b.shape[n]:
+                entries[n] = I[n] // (a.shape[n] // b.shape[n])
+            if a.shape[n] < b.shape[n]:
+                entries[n] = I[n] * (b.shape[n] // a.shape[n])
```
Comment (strongoier) [inline python/taichi/lang/snode.py:410]:
```suggestion
        entries = [I[i] for i in range(n)]
        for i in impl.static(range(min(n, min(len(a.shape), len(b.shape))))):
            if a.shape[i] > b.shape[i]:
                entries[i] = I[i] // (a.shape[i] // b.shape[i])
            if a.shape[i] < b.shape[i]:
                entries[i] = I[i] * (b.shape[i] // a.shape[i])
```

**Review by strongoier — APPROVED** [review]
Let's avoid calling `matrix.Vector` in the list case under `real_matrix=True`.

**Pull request merged**